services/finetuning/tsfmfinetuning/finetuning.py

```diff
@@ -0,0 +1,243 @@
+# Copyright contributors to the TSFM project
+#
+"""Tsfmfinetuning Runtime"""
+
+import logging
+from pathlib import Path
+from typing import Any, Dict
+
+import pandas as pd
+from fastapi import APIRouter, HTTPException
+from transformers import EarlyStoppingCallback, Trainer, TrainingArguments, set_seed
+
+from tsfm_public import TimeSeriesPreprocessor
+from tsfm_public.toolkit.dataset import ForecastDFDataset
+from tsfm_public.toolkit.util import select_by_fixed_fraction
+
+from . import TSFM_ALLOW_LOAD_FROM_HF_HUB
+from .constants import API_VERSION
+from .ftpayloads import (
+    AsyncCallReturn,
+    BaseTuneInput,
+    TinyTimeMixerForecastingTuneInput,
+    TuneTypeEnum,
+)
+from .hfutil import load_config, load_model, register_config
+from .ioutils import to_pandas
+
+
+LOGGER = logging.getLogger(__file__)
+
+
+class FinetuningRuntime:
+    def __init__(self, config: Dict[str, Any] = {}):
+        self.config = config
+        model_map = {}
+
+        if "custom_modules" in config:
+            for custom_module in config["custom_modules"]:
+                register_config(
+                    custom_module["model_type"],
+                    custom_module["model_config_name"],
+                    custom_module["module_path"],
+                )
+                LOGGER.info(f"registered {custom_module['model_type']}")
+
+                model_map[custom_module["model_config_name"]] = custom_module["module_path"]
+
+        self.model_to_module_map = model_map
+
+    def add_routes(self, app):
+        self.router = APIRouter(prefix=f"/{API_VERSION}/finetuning", tags=["finetuning"])
+        self.router.add_api_route(
+            "/tinytimemixer/forecasting",
+            self.finetuning,
+            methods=["POST"],
+            response_model=AsyncCallReturn,
+        )
+        app.include_router(self.router)
+
+    def load(self, model_path: str):
+        try:
+            preprocessor = TimeSeriesPreprocessor.from_pretrained(model_path)
+            LOGGER.info("Successfully loaded preprocessor")
+        except OSError:
+            preprocessor = None
+            LOGGER.info("No preprocessor found")
+
+        # load config and model
+        conf = load_config(
+            model_path,
+        )
+
+        model = load_model(
+            model_path,
+            config=conf,
+            module_path=self.model_to_module_map.get(conf.__class__.__name__, None),
+        )
+        LOGGER.info("Successfully loaded model")
+        return model, preprocessor
+
+    def finetuning(self, input: TinyTimeMixerForecastingTuneInput, tuned_model_name: str, output_dir: Path):
+        try:
+            LOGGER.info("calling forecast_common")
```

Comment:
finetuning_common?

Reply:
It was there originally but it seems you were pushing more functionality under the runtime class including the references to the self.config. I would actually prefer to separate the service route definitions from underlying implementation so we can refactor.